### ISSUE TITLE
gpuarray.gpudata.as_buffer is not writable in Python 3

### ISSUE THREAD

**Issue by leofang** [inducer/pycuda#182]
Environment:
Python 3.6.5
mpi4py v3.0.0 installed against Open MPI v3.0.1 (--with-cuda)
pycuda v2017.1.1

MWE:
```Python
#!/usr/bin/env python
'''
code snippet borrowed from https://gist.github.com/lebedov/8514d3456a94a6c73e6d (slightly modified)
run with "mpirun -n 2 python test.py"
'''
import atexit
import sys

# PyCUDA 2014.1 and later have built-in support for wrapping GPU memory with a
# buffer interface:
import pycuda
if pycuda.VERSION >= (2014, 1):
    bufint = lambda arr: arr.gpudata.as_buffer(arr.nbytes)
else:
    import cffi
    ffi = cffi.FFI()
    bufint = lambda arr: ffi.buffer(ffi.cast('void *', arr.ptr), arr.nbytes)

import numpy as np
from mpi4py import MPI

import pycuda.driver as drv
import pycuda.gpuarray as gpuarray
drv.init()

def dtype_to_mpi(t):
    if hasattr(MPI, '_typedict'):
        mpi_type = MPI._typedict[np.dtype(t).char]
    elif hasattr(MPI, '__TypeDict__'):
        mpi_type = MPI.__TypeDict__[np.dtype(t).char]
    else:
        raise ValueError('cannot convert type')
    return mpi_type

comm = MPI.COMM_WORLD
size = comm.Get_size()
rank = comm.Get_rank()

N_gpu = drv.Device(0).count()
if N_gpu < 2:
    sys.stdout.write('at least 2 GPUs required')
else:
    dev = drv.Device(rank)
    ctx = dev.make_context()
    atexit.register(ctx.pop)
    atexit.register(MPI.Finalize)

    if rank == 0:
        x_gpu = gpuarray.arange(100, 200, 10, dtype=np.double)
        print ('before ({0}): {1}'.format(rank, x_gpu))
        comm.Send([bufint(x_gpu), dtype_to_mpi(x_gpu.dtype)], dest=1)
        print ('sent')
        print ('after  ({0}): {1}'.format(rank, x_gpu))
    elif rank == 1:
        x_gpu = gpuarray.zeros(10, dtype=np.double)
        print ('before ({0}): {1}'.format(rank, x_gpu))
        comm.Recv([bufint(x_gpu), dtype_to_mpi(x_gpu.dtype)], source=0)
        print ('received')
        print ('after  (%i): ' % rank)+str(x_gpu)
```

Output: 
```
before (1): [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
before (0): [100. 110. 120. 130. 140. 150. 160. 170. 180. 190.]
sent
after  (0): [100. 110. 120. 130. 140. 150. 160. 170. 180. 190.]
Traceback (most recent call last):
  File "mpi4py_pycuda_test.py", line 63, in <module>
    comm.Recv([bufint(x_gpu), dtype_to_mpi(x_gpu.dtype)], source=0)
  File "mpi4py/MPI/Comm.pyx", line 283, in mpi4py.MPI.Comm.Recv
  File "mpi4py/MPI/msgbuffer.pxi", line 402, in mpi4py.MPI.message_p2p_recv
  File "mpi4py/MPI/msgbuffer.pxi", line 388, in mpi4py.MPI._p_msg_p2p.for_recv
  File "mpi4py/MPI/msgbuffer.pxi", line 155, in mpi4py.MPI.message_simple
  File "mpi4py/MPI/msgbuffer.pxi", line 94, in mpi4py.MPI.message_basic
  File "mpi4py/MPI/asbuffer.pxi", line 308, in mpi4py.MPI.getbuffer
  File "mpi4py/MPI/asbuffer.pxi", line 135, in mpi4py.MPI.PyMPI_GetBuffer
BufferError: memoryview: underlying buffer is not writable
-------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
-------------------------------------------------------
--------------------------------------------------------------------------
mpirun detected that one or more processes exited with non-zero status, thus causing
the job to be terminated. The first process to do so was:

  Process name: [[63172,1],1]
  Exit code:    1
--------------------------------------------------------------------------
```

Notes:
1. See also recent mailing list: https://www.mail-archive.com/pycuda@tiker.net/msg03900.html
2. the buffer is there (and can be sent), but it can't be written.

**Comment by leofang** [comment]
My bad, commit 1594f550c2b680f32e4678f4ff3bc67268f93fc1 has already addressed this issue. Thanks.

**Comment by inducer** [comment]
Whoa. I had somehow missed that the PyCUDA releases have fallen that far behind master again. Thanks for the (well-written!) bug report, even if I could have saved you the trouble. (Sorry!)

I've just released 2018.1 off of the latest code base.
